Update Firefox desktop pages post GitHub migration (#284)
flodolo/documentation: src/products/firefox_desktop/fluent_migrations.md

[Fluent migrations](https://firefox-source-docs.mozilla.org/l10n/migrations/index.html) are used to migrate existing translations from legacy formats (DTD, properties) to Fluent, or existing Fluent messages to new messages or different FTL files.

At the core of each migration there is a Python file, called a **migration recipe**, that instructs the system on how to port existing translations to Fluent. Typically, a patch migrating content to Fluent will:
* Add new Fluent strings or files.
* Remove obsolete strings.
* Include a migration recipe.

> Once one such patch lands in [gecko-dev](https://github.com/mozilla/gecko-dev), string changes will be pushed to a quarantine repository (the `update` branch of [firefox-l10n-source](https://github.com/mozilla-l10n/firefox-l10n-source)). At this point, this new content is not visible yet in [Pontoon](https://pontoon.mozilla.org/).

Reviewer: ```suggestion
Once one such patch lands in [gecko-dev](https://github.com/mozilla/gecko-dev), string changes will be pushed to the `update` branch of [firefox-l10n-source](https://github.com/mozilla-l10n/firefox-l10n-source). At this point, this new content is not visible yet in [Pontoon](https://pontoon.mozilla.org/).
```

It's not a quarantine repository anymore, in case a "quarantine area".

The goal of a migration is to port existing translations to a new format or file without forcing the community localizers to retranslate them. For this to happen, migrations need to be performed before the new strings are pushed to the `main` branch of the `firefox-l10n-source` repository and become available in Pontoon.

> This means that patches that include migrations should only be pushed from the quarantine repository to `main` following the steps described in this document.

Reviewer: ```suggestion
This means that patches that include migrations should only be pushed from the `update` branch to `main` following the steps described in this document.
```

## Running Fluent migrations

This section of the document describes in detail which operations need to be performed in order to run migrations.

### Prerequisites

Several access rights are required:
* Admin access to `firefox-l10n-source` and `firefox-l10n` (as a member of the `l10n-firefox` team).
* Admin access to [Pontoon](https://pontoon.mozilla.org/).
* An account on Heroku with access to [Pontoon apps](https://dashboard.heroku.com/apps/mozilla-pontoon/resources).

In order to run migrations:
* Install the [fluent.migrate](https://pypi.org/project/fluent.migrate/) Python package.
* Clone the [firefox-l10n](https://github.com/mozilla-l10n/firefox-l10n) repository.
* Clone the [firefox-l10n-source](https://github.com/mozilla-l10n/firefox-l10n-source) repository.

### Ensure there are no blocking issues in the update branch of firefox-l10n-source

> As part of migrations, content needs to be pushed from the `update` branch to the `main` branch of `firefox-l10n-source`. Since content can’t be pushed selectively, any pending issue needs to be solved before starting the process (eventually backing out problematic changes). For more information about the review process, see [this document](review.md).

Reviewer: This is not true at this point. You can make manual changes to the `update` branch to take out issues.

> As part of migrations, content needs to be pushed from the `update` branch to the `main` branch of `firefox-l10n-source`. Any pending issue needs to be solved before starting the process (eventually backing out problematic changes) or need to be temporarily removed manually to prevent problematic strings being exposed to localizers. For more information about the review process, see [this document](review.md).

Reviewer: ```suggestion
As part of migrations, content needs to be pushed from the `update` branch to the `main` branch of `firefox-l10n-source`. Any pending issue needs to be solved before starting the process (eventually manually backing out problematic changes) to prevent problematic strings being exposed to localizers. For more information about the review process, see [this document](review.md).
```

A series of tools and helpers to run migrations is available in [this repository](https://github.com/flodolo/fluent-migrations):
* Clone the repository locally and follow the instructions available in the [README](https://github.com/flodolo/fluent-migrations/blob/master/README.md) to set the configuration file.
* Add the new migration recipes in the `recipes` folder. The script will look for any Python file starting with `bug_` in that folder, so it’s possible to run multiple migrations in one pass. More information about the folder’s structure is available in the repository’s README.

At this point, while each migration has been tested as part of the review before landing, it’s always good to run the migration against one locale, e.g. `it`:

> * Update the locale’s repository, check the results of `compare-locales`. Assuming the l10n repository is cloned in `~/migrations/firefox-l10n`, and the quarantine repository is in `~/migrations/firefox-l10n-source` and the `update` branch is checked out, the command to run is `compare-locales ~/migrations/firefox-l10n-source/_configs/browser.toml ~/migrations/firefox-l10n it`. Save the output in order to compare it with the results after the migration.

Reviewer: ```suggestion
* Update the locale’s repository, check the results of `compare-locales`. Assuming the l10n repository is cloned in `~/migrations/firefox-l10n`, and the quarantine repository is in `~/migrations/firefox-l10n-source` with the `update` branch checked out, the command to run is `compare-locales ~/migrations/firefox-l10n-source/_configs/browser.toml ~/migrations/firefox-l10n it`. Save the output in order to compare it with the results after the migration.
```

* Update the locale’s repository, check the results of `compare-locales`. Assuming l10n repositories are cloned in `~/migrations/locales`, and the quarantine repository in `~/migrations/gecko-strings-quarantine`, the command to run is `compare-locales ~/migrations/gecko-strings-quarantine/_configs/browser.toml ~/migrations/locales it`. Save the output in order to compare it with the results after the migration.
* Update the locale’s repository, check the results of `compare-locales`. Assuming the l10n repository is cloned in `~/migrations/firefox-l10n`, and the quarantine repository is in `~/migrations/firefox-l10n-source` and the `update` branch is checked out, the command to run is `compare-locales ~/migrations/firefox-l10n-source/_configs/browser.toml ~/migrations/firefox-l10n it`. Save the output in order to compare it with the results after the migration.
* Run the migration, without pushing, only for `it`: `./scripts/migrate it wet-run`. Then run compare-locales again, and check if the results are as expected: migrated strings should not appear as missing anymore, and there should be no errors.

> * If everything looks as expected, re-run the script adding the push option `./scripts/migrate it wet-run push`.

Reviewer: Why removing this? If something goes wrong in the next step, you'll have a pending commit locally.

Author: I was trying to reduce the amount of testing steps and simplify instructions, this seemed redundant since it'd be run again when you run all (and running everything doesn't take as much time anymore). I'm not quite sure what the problem with a pending local commit is, if the step failed here there'd still be a pending commit either way.

Reviewer: Right, I think we can get away without this step, especially considering that the default is running migrations in a branch. If you were running against main, you risk having to pull and rebase, but this was much more disruptive on hg (there's no "pull and rebase", there's "pull+merge+commit+push" and you might not notice the issue when running 100+ locales).

Before [merging updates](review.md#merge-reviewed-strings-to-main-branch-of-firefox-l10n-source) to the `main` branch of `firefox-l10n-source`, make sure that:

> * The new strings and migrations are available in the `update` branch.

Reviewer: Migrations are not in firefox-l10n-source. What did you mean here?

Author: I've added some clarifying language to avoid confusion. This is the step to merge source strings (i.e. `update` into `main` on `firefox-l10n-source`).
